python/ray/serve/_private/router.py

```diff
@@ -213,20 +189,103 @@ def _get_aggregated_requests(self):
+            "queued_requests": self.num_queued_requests,
             "running_requests": running_requests,
         }
 
     def process_finished_request(self, replica_tag, *args):
         with self._queries_lock:
             self.num_requests_sent_to_replicas[replica_tag] -= 1
 
+    def shutdown(self):
+        """Shutdown metrics manager gracefully.
+
+        The metrics_pusher needs to be shutdown separately.
+        """
+
+        if self.metrics_pusher:
+            self.metrics_pusher.shutdown()
+
+
+class Router:
+    def __init__(
+        self,
+        controller_handle: ActorHandle,
+        deployment_id: DeploymentID,
+        handle_id: str,
+        self_node_id: str,
+        self_actor_id: str,
+        self_availability_zone: Optional[str],
+        event_loop: asyncio.BaseEventLoop = None,
+        _prefer_local_node_routing: bool = False,
+        _router_cls: Optional[str] = None,
+    ):
+        """Used to assign requests to downstream replicas for a deployment.
+
+        The scheduling behavior is delegated to a ReplicaScheduler; this is a thin
+        wrapper that adds metrics and logging.
+        """
+        self._event_loop = event_loop
+        self.deployment_id = deployment_id
+
+        if _router_cls:
+            self._replica_scheduler = load_class(_router_cls)(
+                event_loop=event_loop, deployment_id=deployment_id
+            )
+        else:
+            self._replica_scheduler = PowerOfTwoChoicesReplicaScheduler(
+                event_loop,
+                deployment_id,
+                _prefer_local_node_routing,
+                RAY_SERVE_PROXY_PREFER_LOCAL_AZ_ROUTING,
+                self_node_id,
+                self_actor_id,
+                self_availability_zone,
+            )
+
+        logger.info(
+            f"Using router {self._replica_scheduler.__class__}.",
+            extra={"log_to_stderr": False},
+        )
+
+        self.long_poll_client = LongPollClient(
+            controller_handle,
+            {
+                (
+                    LongPollNamespace.RUNNING_REPLICAS,
+                    deployment_id,
+                ): self.update_running_replicas,
+                (
+                    LongPollNamespace.AUTOSCALING_CONFIG,
+                    deployment_id,
+                ): self.update_autoscaling_config,
+            },
+            call_in_event_loop=event_loop,
+        )
+
+        # For autoscaling deployments.
+        self._metrics_manager = RouterMetricsManager(
+            deployment_id, handle_id, controller_handle
+        )
+
+    def update_running_replicas(self, running_replicas: List[RunningReplicaInfo]):
+        self._replica_scheduler.update_running_replicas(running_replicas)
+        self._metrics_manager.update_running_replicas(running_replicas)
+
+    def update_autoscaling_config(self, autoscaling_config: AutoscalingConfig):
+        self._metrics_manager.update_autoscaling_config(
```

Comment:
Hmm I know this is mostly just moving code around and not sure if this is already discussed in other forums. But maybe the some of logics for `update_autoscaling_config` should stay in here? And only when there is a change in the autoscaling config it calls the metrics manager to start/ stop metrics pusher? Just feel a bit weird to be the metrics manager needs to also deal and keeping a copy of `autoscaling_config`

Reply:
Hmmm, could you explain more why you think the router should deal with the autoscaling config instead? Autoscaling config is only used for metrics aggregation (`look_back_period_s`) and metrics pushing (`metrics_interval_s`) in the router right now, which is why I felt it makes sense to move it completely into the metrics manager.